srml/offences/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+
+//! # Offences Module
+//!
+//! Tracks reported offences
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod mock;
+mod tests;
+
+use rstd::vec::Vec;
+use support::{
+	StorageDoubleMap, decl_module, decl_event, decl_storage, Parameter,
+};
+use sr_primitives::Perbill;
+use sr_staking_primitives::{
+	SessionIndex,
+	offence::{Offence, ReportOffence, TimeSlot, Kind, OnOffenceHandler, OffenceDetails},
+};
+
+/// Offences trait
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Full identification of the validator.
+	type IdentificationTuple: Parameter;
+	/// A handler called for every offence report.
+	type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Offences {
+		/// A mapping between unique `TimeSlots` within a particular session and the offence `Kind` into
+		/// a vector of offending authorities.
+		///
+		/// It's important that we store all authorities reported for an offence for any kind and
+		/// timeslot since the slashing will increase based on the length of this vec.
+		OffenceReports get(offence_reports):
+			double_map Kind, blake2_256((SessionIndex, TimeSlot))
+			=> Vec<OffenceDetails<T::AccountId, T::IdentificationTuple>>;
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		/// There is an offence reported of the given `kind` happened at the `session_index` and
+		/// (kind-specific) time slot. This event is not deposited for duplicate slashes.
+		Offence(Kind, SessionIndex, TimeSlot),
+	}
+);
+
+decl_module! {
+	/// Offences module, currently just responsible for taking offence reports.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T::IdentificationTuple, O>
+	for Module<T> where
+	T::IdentificationTuple: Clone,
+{
+	fn report_offence(reporter: Option<T::AccountId>, offence: O) {
+		let offenders = offence.offenders();
+		let time_slot = offence.time_slot();
+		let session = offence.session_index();
+		let validator_set_count = offence.validator_set_count();
+
+		// Check if an offence is already reported for the offender authorities
+		// and otherwise stores that report.
+		let mut changed = false;
+		let all_offenders = <OffenceReports<T>>
+			::mutate(&O::ID, &(session, time_slot), |offending_authorities| {
+				for offender in offenders {
+					// TODO [slashing] This prevents slashing for multiple reports of the same kind at the same slot,
+					// note however that we might do that in the future if the reports are not exactly the same (dups).
+					// De-duplication of reports is tricky though, we need a canonical form of the report
+					// (for instance babe equivocation can have headers swapped).
```

Comment:
well, this isn't actually difficult.

`footprint = babe_equivoation_footprint(who, slot_number)`.

Reply:
Does this function exist already? What if you issue 4 different headers in the same slot? I guess you should be slashed for all possible pairs, no?

Reply:
@tomusdrw 

We've decided not to slash for every equivocation - only the first. The reason being that you can equivocate an infinite amount of times for relatively low practical cost, and then you would easily overwhelm the chain processing all the reports.